pkg/controller/baremetalhost/baremetalhost_controller.go

```diff
@@ -198,54 +198,54 @@ func (r *ReconcileBareMetalHost) Reconcile(request reconcile.Request) (result re
 		return result, err
 	}
 
-	// Check for a "discovered" host vs. one that we have all the info for.
-	if host.Spec.BMC.Address == "" {
-		reqLogger.Info(bmc.MissingAddressMsg)
-		dirty := host.SetOperationalStatus(metalkubev1alpha1.OperationalStatusDiscovered)
-		if dirty {
-			err = r.saveStatus(host)
-			if err != nil {
+	// Retrieve the BMC details from the host spec and validate host
+	// BMC details and build the credentials for talking to the
+	// management controller.
+	bmcCreds, bmcCredsSecret, err := r.buildAndValidateBMCCredentials(request, host)
+	if err != nil {
+		switch err.(type) {
+		// We key on empty bmc address and empty bmc credentials fields as a
+		// trigger the host needs to be put into a discovered status and requeue it for
+		// later.  We do set an error message (but not an error state) on the host
+		// so we can understand what we may be waiting on.
+		case *EmptyBMCAddressError, *EmptyBMCSecretError:
+			dirty := host.SetOperationalStatus(metalkubev1alpha1.OperationalStatusDiscovered)
+			if dirty {
+				// Set the host error message directly
+				// as we cannot use SetErrorCondition which
+				// overwrites our discovered state
+				host.Status.ErrorMessage = err.Error()
+				saveErr := r.saveStatus(host)
+				if saveErr != nil {
+					return reconcile.Result{Requeue: true}, saveErr
+				}
 				// Only publish the event if we do not have an error
 				// after saving so that we only publish one time.
 				r.publishEvent(request,
-					host.NewEvent("Discovered", "Discovered host without BMC address"))
+					host.NewEvent("Discovered", fmt.Sprintf("Discovered host with unusable BMC details: %s", err.Error())))
 			}
-			// Without the address we can't do any more so we return here
-			// without checking for an error.
-			return reconcile.Result{Requeue: true}, err
-		}
-		reqLogger.Info("nothing to do for discovered host without BMC address")
-		return reconcile.Result{}, nil
-	}
-	if host.Spec.BMC.CredentialsName == "" {
-		reqLogger.Info(bmc.MissingCredentialsMsg)
-		dirty := host.SetOperationalStatus(metalkubev1alpha1.OperationalStatusDiscovered)
-		if dirty {
-			err = r.saveStatus(host)
-			if err != nil {
-				// Only publish the event if we do not have an error
-				// after saving so that we only publish one time.
-				r.publishEvent(request,
-					host.NewEvent("Discovered", "Discovered host without BMC credentials"))
+			return reconcile.Result{Requeue: true, RequeueAfter: hostErrorRetryDelay}, nil
+		// In the event a credential secret is defined, but we cannot find it, or the secret is
+		// missing the required fields, or the BMC address is defined but malformed we set the
+		// host into an error state and requeue it for evaluation later
```

Comment:
The only one of those errors that requires requeueing is the secret not being found.

If the secret exists but is missing any data or has bad data, we will have set the host as its owner and that will cause the host to be reconciled when the secret contents are edited to correct the problem.

If the BMC address is bad in some way, fixing it will require editing the host spec, which will trigger reconciliation.

A missing secret might mean a typo in the name, or it might mean the secret really doesn't exist. We can't know the difference, and in the latter case the user might "fix" the problem by creating the secret but that won't trigger reconciliation for the host because we didn't set up the owner relationship. So, if we have a secret name but can't load it at all, then we need to schedule the host to be reconciled again.

Reply:
Fixed